rasterio/profiles.py

```diff
@@ -1,41 +1,69 @@
 """Raster dataset profiles."""
 
+import warnings
+
+from rasterio.compat import UserDict
 from rasterio.dtypes import uint8
 
 
-class Profile:
+class Profile(UserDict):
     """Base class for Rasterio dataset profiles.
 
-    Subclasses will declare a format driver and driver-specific
-    creation options.
+    Subclasses will declare driver-specific creation options.
     """
 
-    driver = None
     defaults = {}
 
-    def __call__(self, **kwargs):
-        """Return a mapping of keyword args for writing a new datasets.
+    def __init__(self, data={}, **kwds):
+        """Create a new profile based on the class defaults, which are
+        overlaid with items from the `data` dict and keyword arguments."""
+        UserDict.__init__(self)
+        initdata = self.defaults.copy()
+        initdata.update(data)
+        initdata.update(**kwds)
+
+        if 'affine' in initdata and 'transform' in initdata:
+            warnings.warn("affine item is deprecated, use transform only",
+                          DeprecationWarning)
+            del initdata['affine']
+        elif 'affine' in initdata:
+            warnings.warn("affine item is deprecated, use transform instead",
+                          DeprecationWarning)
+            initdata['transform'] = initdata.pop('affine')
+
+        self.data.update(initdata)
+
+    def __getitem__(self, key):
+        """Like normal item access but with affine alias."""
+        if key == 'affine':
+            key = 'transform'
+            warnings.warn("affine item is deprecated, use transform instead",
+                          DeprecationWarning)
+        return self.data[key]
+
+    def __setitem__(self, key, val):
+        """Like normal item setter but forbidding affine item."""
+        if key == 'affine':
+            raise TypeError("affine key is prohibited")
```

Comment:
Would `AttributeError` be a better fit here? The type of key (string) isn't the source of the problem so `TypeError` feels imprecise. Not a blocker, we can assess this later.